### PR TITLE
CRFEntityExtractor supports composite entities

### PR DESCRIPTION
**Pull request by tabergma**
**Proposed changes**:
- Update `CRFEntityExtractor` to support composite entities.
- Unify way of handling BILOU tagging schema in `CRFEntityExtractor` and `DIETClassifier`

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [x] updated the documentation
- [x] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/master/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

### CODE REVIEW

**Comment by tabergma** [comment]
Just as a note: My plan is to implement a check for consistent entity annotations in the training data and cleaning up inconsistent BILOU predictions in a separate PR. Both checks should happen always, independent from the extractor you use.

File: rasa/nlu/extractors/crf_entity_extractor.py [hunk 1]
```diff
@@ -130,216 +150,142 @@ def train(
...
+        """Add predicted entity tags to CRF tokens."""
+        if ENTITY_ATTRIBUTE_TYPE in predictions:
+            _tags, _ = self._most_likely_tag(predictions[ENTITY_ATTRIBUTE_TYPE])
+            for tag, token in zip(_tags, crf_tokens):
+                token.entity_tag = tag
+
+    def _most_likely_tag(
+        self, predictions: List[Dict[Text, float]]
+    ) -> Tuple[List[Text], List[float]]:
+        """Get the entity tags with the highest confidence.
```
Comment (Ghostvv) [inline rasa/nlu/extractors/crf_entity_extractor.py:235]: in crf tags are found through the recurrent process that maximizes total sequence confidence. I thought, it is part of the crf

File: rasa/nlu/extractors/crf_entity_extractor.py [hunk 2]
```diff
@@ -130,216 +150,142 @@ def train(
...
-    ):
-        # using the BILOU tagging scheme
-        json_ents = []
-        word_idx = 0
-        while word_idx < len(tokens):
-            end_idx, confidence, entity_label = self._handle_bilou_label(
-                word_idx, entities
-            )
+            if self.component_config["BILOU_flag"]:
+                bilou_utils.ensure_consistent_bilou_tagging(_tags)
```
Comment (Ghostvv) [inline rasa/nlu/extractors/crf_entity_extractor.py:282]: doing this in DIET will be separate PR?

File: rasa/nlu/extractors/extractor.py [hunk 3]
```diff
@@ -333,3 +346,149 @@ def filter_trainable_entities(
             )
 
         return filtered
+
+    def create_entities(
```
Comment (Ghostvv) [inline rasa/nlu/extractors/extractor.py:350]: I don't think this name the best describes what it does

File: rasa/nlu/utils/bilou_utils.py [hunk 4]
```diff
@@ -240,22 +244,75 @@ def _add_bilou_tags_to_entities(
...
-def _get_entity_positions(entities: List[Tuple[int, int, Text]]) -> Set[int]:
-    entity_positions = set()
+def ensure_consistent_bilou_tagging(predicted_tags: List[Text]) -> List[Text]:
+    """
+    Ensure predicted tags follow the BILOU tagging schema.
 
-    for start_pos, end_pos, label in entities:
-        for i in range(start_pos, end_pos):
-            entity_positions.add(i)
+    We assume that starting B- tags are correct. Followed tags that belong to start
```
Comment (Ghostvv) [inline rasa/nlu/utils/bilou_utils.py:251]: this should also be represented in confidences then, like we should remove the confidence for tags that were substituted. Should it rather be majority vote, like if all the tags except B-tag is the same?

File: rasa/nlu/extractors/crf_entity_extractor.py (hunk 1 above)
Comment (tabergma) [reply rasa/nlu/extractors/crf_entity_extractor.py:235]: The output of the CRF is a ranking of tags, so we get all possible tags with confidence values. This function just returns the tag with the highest confidence.

File: rasa/nlu/extractors/crf_entity_extractor.py (hunk 2 above)
Comment (tabergma) [reply rasa/nlu/extractors/crf_entity_extractor.py:282]: I added a similar line to DIET in this PR.

File: rasa/nlu/utils/bilou_utils.py (hunk 4 above)
Comment (tabergma) [reply rasa/nlu/utils/bilou_utils.py:251]: If we have an entity that covers multiple words, we take the lowest confidence score among those words. We could leave out the "misclassified" tag in this calculation.
I did not want to update the logic for now. Was thinking of using majority vote, but we might want to change this in a separate PR.

File: rasa/nlu/extractors/extractor.py [hunk 5]
```diff
@@ -333,3 +346,149 @@ def filter_trainable_entities(
...
+        self,
+        text: Text,
+        tokens: List[Token],
+        tags: Dict[Text, List[Text]],
+        confidences: Optional[Dict[Text, List[float]]] = None,
+    ) -> List[Dict[Text, Any]]:
+        """
+        Converts predicted tags into proper entities.
```
[the remark this answers is not in the document]
Comment (ricwo) [reply rasa/nlu/extractors/extractor.py:358]: what do you mean by proper entities?

File: rasa/nlu/utils/bilou_utils.py [hunk 6]
```diff
@@ -240,22 +244,72 @@ def _add_bilou_tags_to_entities(
...
+        tag = tag_without_prefix(predicted_tag)
+
+        if not prefix:
+            if entity_found:
+                logger.debug(
+                    f"Inconsistent BILOU tagging schema found. Missing "
+                    f"'L-{last_tag}'."
+                )
+
+            entity_found = False
```
Comment (ricwo) [inline rasa/nlu/utils/bilou_utils.py]: this can move inside the `if entity_found:` case

File: rasa/nlu/utils/bilou_utils.py [hunk 7]
```diff
@@ -240,22 +244,72 @@ def _add_bilou_tags_to_entities(
...
+            last_tag = NO_ENTITY_TAG
+            continue
+
+        if prefix == "B":
```
Comment (ricwo) [inline rasa/nlu/utils/bilou_utils.py:267]: why are these prefixes different from the ones listed in `BILOU_PREFIXES` (without the `-` suffix)?

File: tests/nlu/extractors/test_crf_entity_extractor.py [hunk 8]
```diff
@@ -1,257 +1,116 @@
...
-                ["low", "title", "upper", "pos", "pos2"],
-            ]
-        }
-    )
-    tokenizer = SpacyTokenizer()
-
-    training_data = TrainingData(training_examples=examples)
-    tokenizer.train(training_data)
-    extractor.train(training_data)
+def as_pipeline(*components):
```
Comment (ricwo) [inline tests/nlu/extractors/test_crf_entity_extractor.py:15]: types please

File: tests/nlu/extractors/test_crf_entity_extractor.py [hunk 9]
```diff
@@ -1,257 +1,116 @@
...
+    return [{"name": c} for c in components]
 
-    sentence = "italian restaurant"
-    message = Message(sentence, {SPACY_DOCS[TEXT]: spacy_nlp(sentence)})
 
-    tokenizer.process(message)
-    extractor.process(message)
+async def test_train_persist_load_with_composite_entities(component_builder, tmpdir):
```
Comment (ricwo) [inline tests/nlu/extractors/test_crf_entity_extractor.py:19]: types

File: tests/nlu/extractors/test_crf_entity_extractor.py [hunk 10]
```diff
@@ -1,257 +1,116 @@
...
+                    ],
+                    ["low", "title", "upper", "pos", "pos2"],
+                ],
+                "BILOU_flag": True,
+            }
+        ),
+    ],
+)
+async def test_train_persist_with_different_configurations(
+    config_params, component_builder, tmpdir
```
Comment (ricwo) [inline tests/nlu/extractors/test_crf_entity_extractor.py:85]: types please

File: tests/nlu/extractors/test_crf_entity_extractor.py [hunk 11]
```diff
@@ -1,257 +1,116 @@
...
-                    "value": entity["value"],
-                    "entity": entity["entity"],
-                    "confidence": 0.8,
-                }
+    assert len(detected_entities) == 1
+    assert detected_entities[0]["entity"] == "cuisine"
+    assert detected_entities[0]["value"] == "italian"
 
 
 def test_crf_use_dense_features(spacy_nlp):
```
Comment (ricwo) [inline tests/nlu/extractors/test_crf_entity_extractor.py:116]: types please

File: tests/nlu/utils/test_bilou_utils.py [hunk 12]
```diff
@@ -139,3 +142,43 @@ def test_apply_bilou_schema():
...
+            "Found 'L-person', but last seen 'location'",
+        ),
+        (
+            ["O", "U-person", "B-location", "U-person", "O", "U-person", "O"],
+            "Missing 'L-location'",
+        ),
+    ],
+)
+def test_check_consistent_bilou_tagging(
+    tags: List[Text], message: Optional[Text], caplog
```
Comment (ricwo) [inline tests/nlu/utils/test_bilou_utils.py]:
```suggestion
    tags: List[Text], message: Optional[Text], caplog: LogCaptureFixture
```
with module-level
```python
from _pytest.logging import LogCaptureFixture
```

File: tests/nlu/utils/test_bilou_utils.py [hunk 13]
```diff
@@ -139,3 +142,43 @@ def test_apply_bilou_schema():
...
+):
+    with caplog.at_level(logging.DEBUG):
```
Comment (ricwo) [inline tests/nlu/utils/test_bilou_utils.py:172]: I don't think this is needed, as the log itself is already a `DEBUG`

File: tests/nlu/utils/test_bilou_utils.py [hunk 14]
```diff
@@ -139,3 +142,43 @@ def test_apply_bilou_schema():
...
+        bilou_utils.check_consistent_bilou_tagging(tags)
```
Comment (ricwo) [inline tests/nlu/utils/test_bilou_utils.py]:
```suggestion
    bilou_utils.check_consistent_bilou_tagging(tags)
```

File: tests/nlu/extractors/test_crf_entity_extractor.py (hunk 8 above)
Comment (ricwo) [reply tests/nlu/extractors/test_crf_entity_extractor.py:15]: maybe `pipeline_from_components`?

File: rasa/nlu/utils/bilou_utils.py (hunk 7 above)
Comment (ricwo) [reply rasa/nlu/utils/bilou_utils.py:267]: what do you think about creating a `NamedTuple` and accessing the prefixes with `nt.beginning`, `nt.inside` etc? this could replace the `BILOU_PREFIXES` at the top and be used across this file

File: tests/nlu/extractors/test_crf_entity_extractor.py [hunk 15]
```diff
@@ -285,12 +144,32 @@ def test_crf_use_dense_features(spacy_nlp):
...
+
+
+@pytest.mark.parametrize(
+    "entity_predictions, expected_label, expected_confidence",
+    [
+        ([{"O": 0.34, "B-person": 0.03, "I-person": 0.85}], ["I-person"], [0.88]),
+        ([{"O": 0.99, "person": 0.03}], ["O"], [0.99]),
+    ],
+)
+def test__most_likely_entity(
```
Comment (ricwo) [inline tests/nlu/extractors/test_crf_entity_extractor.py:165]:
```suggestion
def test_most_likely_entity(
```

File: rasa/nlu/components.py [hunk 16]
```diff
@@ -229,19 +226,17 @@ def validate_required_components_from_data(
         )
 
     if data.entity_examples and not any_components_in_pipeline(
-        {"DIETClassifier"}, pipeline
+        {"DIETClassifier", "CRFEntityExtractor"}, pipeline
```
Comment (ricwo) [inline rasa/nlu/components.py:229]: you mention in the PR that this contains changes for `DIETClassifier` as well - are you still planning to add that to this PR?

File: rasa/nlu/utils/bilou_utils.py (hunk 4 above)
Comment (Ghostvv) [reply rasa/nlu/utils/bilou_utils.py:251]: in this calculation the confidence is basically the confidence of the first entity

File: rasa/nlu/extractors/extractor.py (hunk 3 above)
Comment (tabergma) [reply rasa/nlu/extractors/extractor.py:350]: any suggestion? `create_entities_from_predictions`?

File: examples/formbot/data/nlu.md [hunk 17]
```diff
@@ -1,3 +1,19 @@
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*
+
+- [intent:greet](#intentgreet)
```
[the remark this answers is not in the document]
Comment (tabergma) [reply examples/formbot/data/nlu.md:5]: No idea, those lines are added automatically by some of the tools we are using. I updated this file for local testing and change it back afterwards, but the TOC stayed.

File: rasa/nlu/extractors/crf_entity_extractor.py [hunk 18]
```diff
@@ -352,14 +298,23 @@ def load(
     ) -> "CRFEntityExtractor":
         from sklearn.externals import joblib
 
-        file_name = meta.get("file")
-        model_file = os.path.join(model_dir, file_name)
+        file_names = meta.get("file")
```
[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/nlu/extractors/crf_entity_extractor.py:301]: Yes, not only this breaks the model compatibility. We need to increase the minimum compatible model version in the next release. Don't think we need a warning here, that should be covered with the warning thrown if model is not compatible. But will update the key name.

File: rasa/nlu/extractors/crf_entity_extractor.py [hunk 19]
```diff
@@ -509,45 +437,90 @@ def __get_dense_features(message: Message) -> Optional[List[Any]]:
...
-            c2=self.component_config["L2_c"],
-            # stop earlier
-            max_iterations=self.component_config["max_iterations"],
-            # include transitions that are possible, but not observed
-            all_possible_transitions=True,
-        )
-        self.ent_tagger.fit(X_train, y_train)
+        self.entity_taggers = {}
+
+        for tag_name in self.crf_order:
```
[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/nlu/extractors/crf_entity_extractor.py:554]: we don't have any training progress to log, however, we can log what CRF we are currently training. will add a debug message.

File: rasa/nlu/extractors/extractor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/nlu/extractors/extractor.py]: The JSON structure the model returns in the end.

File: rasa/nlu/extractors/extractor.py [hunk 20]
```diff
@@ -333,3 +346,149 @@ def filter_trainable_entities(
...
+    def tokens_without_cls(message: Message) -> List[Token]:
+        """Return tokens of given message without CLS token.
+
+        Args:
+            message: the message
+
+        Returns:
+            List of tokens.
+        """
+        # [:-1] to remove the CLS token from the list of tokens
```
[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/nlu/extractors/extractor.py:493]: yes, the CLS token is added by all tokenizers to the end of the list of tokens. will add a comment and create a constant.

File: rasa/nlu/extractors/extractor.py (hunk 3 above)
Comment (Ghostvv) [reply rasa/nlu/extractors/extractor.py]: isn't it rather converting predictions into entities? I think `create` here is strange, because at least I associate "creating" differently

File: rasa/nlu/utils/bilou_utils.py [hunk 21]
```diff
@@ -183,39 +192,34 @@ def convert_entity(entity: Dict[Text, Any]) -> Tuple[int, int, Text]:
             entity.get(attribute_key) or NO_ENTITY_TAG,
         )
 
-    return [convert_entity(entity) for entity in message.get(ENTITIES, [])]
+    entities = [convert_entity(entity) for entity in message.get(ENTITIES, [])]
+    # filter out no entity tags
+    return [entity for entity in entities if entity[2] != NO_ENTITY_TAG]
```
[the remark this answers is not in the document]
Comment (tabergma) [reply rasa/nlu/utils/bilou_utils.py:197]: No. `entities` is a list of tuples (start, end, tag value). We want to only return those entities with a tag value != `NO_ENTITY_TAG`. Will add a comment and introduce a variable.

File: rasa/nlu/components.py (hunk 16 above)
Comment (tabergma) [reply rasa/nlu/components.py:229]: changes are already done
